enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+`Domain`. `Nameservers` are responsible for resolving name queries for `Domain`:
+```go
+type Upstream struct {
+	Domain string `json:"domain"`
+	Nameservers []string `json:"nameservers"`
+}
+```
+
+Each instance of CoreDNS performs health checking of `Nameservers`. If `Nameservers` consists of more than one
+nameserver, forwarding will be randomized on first use. When a healthy nameserver returns an error during the exchange,
+the next server in the list is tried. Each server is represented by an IP address or IP address and port if the server
+listens on a UDP port other than 53.
+
+A list of `Upstream` is added to `dnses.opertor.openshift.io`, allowing users the ability to configure multiple
+upstream resolvers to subdomain associations:
+
+```go
+// DNSSpec is the specification of the desired behavior of the DNS.
+type DNSSpec struct {
+	// +optional
+	Upstreams []Upstream `json:"upstreams,omitempty"`
+}
+```
+
+### User Stories
+
+#### Story 1
+
+As a customer with OpenShift running in AWS and connected to my data center by VPN, I want OpenShift DNS to resolve
+name queries for our other internal devices using the DNS servers in our data center.
+
+### Implementation Details/Notes/Constraints
+
+The maximum number of `Nameservers` is 15.
+
+The cluster-dns-operator will generate the `forward` portion of `configmap/dns-default` based on `Upstream` of
+`dnses.operator.openshift.io` instead of `forward` being statically defined. To achieve this:
+
+1. The [default](https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/configmap.yaml) ConfigMap
+asset and associated code from the [manifests pkg](https://github.com/openshift/cluster-dns-operator/blob/master//pkg/manifests/manifests.go#L70:6)
+should be removed.
+
+2. The [desiredDNSConfigMap](https://github.com/openshift/cluster-dns-operator/blob/master/pkg/operator/controller/controller_dns_configmap.go)
+function must be modified to create a ConfigMap type with a `forward` configuration based on `Upstream`.
+If `Upstream` is not present or an invalid `Upstream` is provided, the ConfigMap will contain
+`forward . /etc/resolv.conf`. Otherwise, desiredDNSConfigMap will use the provided `Upstream` to construct the
+forwarding configuration.
+
+#### Updating
+Changes to `Upstream` will trigger a [rolling update](https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/#performing-a-rolling-update)
+of the CoreDNS DaemonSet.
+
+#### Validation
+`Domain` must conform to the [RFC 1123](https://tools.ietf.org/html/rfc1123#page-13) definition of a subdomain. Each
+nameserver in `Nameservers` must be a valid IPv4 or IPv6 address. If the nameserver listens on a UDP port other than 53,
+a valid UDP port number must be specified. A colon is used to separate the address and port, `IP:port` for IPv4 or
+`[IP]:port` for IPv6.
+
+#### Transport
+UDP is used to transport DNS messages and `Upstream` health checks.
```

Comment:
TCP too... usually when the message is too large.

Reply:
> Isn't this unrelated? The resolv.conf governs how the pod resolver works. Here we are configuring the dns server we run that is the first resolver the pod resolver would try. Or did I misunderstand the proposal?

@knobunc the gist of this enhancement is how to manage the CoreDNS `forward` plugin. Although `resolv.conf` is a valid `TO` for the forward plugin, the intent is to prepend nameserevrs using `Upstream` to https://github.com/openshift/cluster-dns-operator/blob/master/assets/dns/configmap.yaml#L15-L17

> TCP too... usually when the message is too large.

I'll update.